quickbudget/views.py

```diff
@@ -0,0 +1,162 @@
+from rest_framework import generics, status, filters
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from quickbudget.api.permissions import (
+    BudgetMembersOnly,
+    AllBudgetExpenseMembersOnly,
+    ExpenseBudgetMembersOnly, IsPostOrIsAuthenticated,
+)
+from quickbudget.api.serializers import (
+    ExpenseSerializer,
+    CategoriesSerializer,
+    BudgetSerializer,
+    MemberAddSerializer,
+    RemoveBudgetMembersSerializer,
+    UserSerializer,
+)
+from quickbudget.models import Budget, Expense, Category, User
+
+
+class ListAddBudgets(generics.ListCreateAPIView):
+    queryset = Budget.objects.all()
+    serializer_class = BudgetSerializer
+
+    def perform_create(self, serializer):
+        # Set the creator field to the current authenticated user
+        serializer.validated_data['creator'] = self.request.user
+
+        if serializer.is_valid():
+            # Create the budget object
+            budget = serializer.save()
+
+            # Add the current user as a member of the budget
+            budget.members.add(self.request.user)
+
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+        return Response(serializer.data, status=status.HTTP_400_BAD_REQUEST)
+
+    def get_queryset(self):
+        return Budget.objects.filter(members=self.request.user)
+
+
+class BudgetDetail(generics.RetrieveUpdateDestroyAPIView):
+    queryset = Budget.objects.all()
+    serializer_class = BudgetSerializer
+    lookup_url_kwarg = "budget_id"
+    permission_classes = [BudgetMembersOnly]
+
+
+class AddExpenses(generics.CreateAPIView):
+    queryset = Expense.objects.all()
+    serializer_class = ExpenseSerializer
+    permission_classes = [AllBudgetExpenseMembersOnly]
+
+
+class ListAddExpenses(generics.ListCreateAPIView):
+    serializer_class = ExpenseSerializer
+    permission_classes = [AllBudgetExpenseMembersOnly]
+    filter_backends = (filters.OrderingFilter,)
+    ordering_fields = ["name", "total", "created_timestamp"]
+
+    def perform_create(self, serializer):
+        # Set the creator field to the current authenticated user
+        serializer.validated_data['author'] = self.request.user
+
+        if serializer.is_valid():
+            # Create the budget object
+            budget = serializer.save()
+
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+        return Response(serializer.data, status=status.HTTP_400_BAD_REQUEST)
+
+    def get_queryset(self):
+        budget = self.kwargs["budget_id"]
+        queryset = Expense.objects.filter(budget_id=budget).order_by(
+            "created_timestamp")
+        category = self.request.query_params.get("category")
+        if category is not None:
```

Comment:
you can replace this code with Ordering settings
https://www.django-rest-framework.org/api-guide/filtering/#specifying-a-default-ordering